model/types.py

```diff
@@ -52,42 +53,74 @@ class MentoExchange(SerializableEnum):
 
 
 class Pair(NamedTuple):
+    """
+    Base Class for Exchange Pairs
+    """
     base: Currency
     quote: Currency
 
+    # pylint: disable = missing-function-docstring;
     def __str__(self):
         return f"{self.base.value}_{self.quote.value}"
+    # pylint: disable = unsubscriptable-object;
+
+    def get_rate(self, prev_state) -> float:
+        market_prices = prev_state['market_price']
+        if self in market_prices:
+            rate = market_prices.get(self)
+        else:
+            pair_1 = [pair for pair in market_prices if pair.base == self.base]
+            assert len(pair_1) == 1, f'No or multiple pairs simulated for {self.base}'
+            pair_1 = pair_1[0]
+            pair_2 = Pair(CryptoAsset.CELO, pair_1.quote)
+            pair_3 = Pair(CryptoAsset.CELO, self.quote)
+            pair_1_2, rate_1_2 = pair_1.implicit_rate(
+                pair_2, prev_state, rate_only=False)
+            rate = pair_1_2.implicit_rate(pair_3, prev_state, self_rate=rate_1_2)
+        return rate
+
+    def implicit_rate(self, other: Pair, prev_state, self_rate=None, other_rate=None,
+                      rate_only: bool = True) -> Union(float, List[Pair, float]):
```

Comment:
And actually you could just add a simple type for clarity:
```python
class Rate(NamedTuple):
  value: float
  pair: Pair
```

so you can have this return `Rate` and if you don't care about the pair you can still do
```python
(value, _) = pair.implicit_rate(...)
```

That way all the functions return `Rate` and I think it's a bit cleaner and consistent, and it's as easy to throw away the pair if you don't need it. Otherwise, you would need to bubble down the `rate_only` param from all the calls to make it useful.